### PR TITLE
[MacOS] Add support for creating and presenting platform views.

### PR DESCRIPTION
**Pull request by RichardJCai**
Add support for:
1. Registering view factories.
2. Creating / Disposing platform views.
3. Presenting platform views in FlutterGLCompositor

Will add follow up PR for static thread merging, there may be visual bugs since raster/ui threads are not merged.

### CODE REVIEW

**Review by cyanglaz — COMMENTED** [review]
We are also missing tests in this PR.

File: shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm
Comment (RichardJCai) [reply shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm]: Yep moved it to the top.

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h
Comment (RichardJCai) [reply shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h:20]: Done

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h
Comment (cyanglaz) [inline shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h:23]: nits: Not sure about my own gramma here :) maybe:
```
Constructing and returning a new NSView instance in this method might cause undefined behavior.
```

**Review by cyanglaz — APPROVED** [review]
LGTM % nits, assuming @iskakaushik LGTMed as well.
Thanks for adding tests!

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h
Comment (RichardJCai) [reply shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h:23]: Oops good catch.

File: shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm
Comment (RichardJCai) [reply shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm:166]: Done

**Comment by iskakaushik** [comment]
Please don't merge this until i've taken a look :-) I think the public API here needs a little more scrutiny.

File: shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm
Comment (cyanglaz) [reply shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm:166]: Add this check outside of the for loop. Also, we should add a DCheck instead of Check to not hard crash production apps.

File: shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm
Comment (cyanglaz) [inline shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm:111]: DCheck here too

File: shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.h
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.h:78]: specified where?

File: shell/platform/darwin/macos/framework/Source/FlutterViewController_Internal.h
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterViewController_Internal.h:19]: call this `platformViewFactories`

File: shell/platform/darwin/macos/framework/Source/FlutterViewController_Internal.h
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterViewController_Internal.h:32]: Create a new class `FlutterPlatformViewController` and move these methods there.

File: shell/platform/darwin/macos/framework/Source/FlutterViewController_Internal.h
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterViewController_Internal.h:45]: parse the args in the caller, the methods here would be better phrased as:

`onCreatePlatformViewWithId:() etc`.

File: shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm:161]: move this method to `FlutterPlatformViewController`.

File: shell/platform/darwin/macos/framework/Source/FlutterViewController_Internal.h
Comment (RichardJCai) [reply shell/platform/darwin/macos/framework/Source/FlutterViewController_Internal.h]: Done

File: shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.h
Comment (RichardJCai) [reply shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.h:81]: Updated to say 
```
  // Remove platform views whos ids are stored in platform_view_controller_'s
  // platformViewsToDispose set.
```

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:570]: In macOS embedding we declare the private methods in the top and add comments.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:620]: this method seems too simple to exist :-)

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:629]: nit: weakSelf

File: shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm:92]: have an equivalent PresentPlatformView method

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViewController.mm
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterPlatformViewController.mm:14]: if (self)

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViewMock.h
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterPlatformViewMock.h:6]: don't need this

Comment (RichardJCai) [reply shell/platform/darwin/macos/framework/Source/FlutterPlatformViewMock.h:6]: Removed

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViewController.mm
Comment (RichardJCai) [reply shell/platform/darwin/macos/framework/Source/FlutterPlatformViewController.mm]: Done

File: shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm
Comment (RichardJCai) [reply shell/platform/darwin/macos/framework/Source/FlutterGLCompositor.mm:92]: Done

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (RichardJCai) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm]: Updated all the other weak_self declarations in FlutterEngine.mm to weakSelf as well for consistency.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (RichardJCai) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:580]: Done

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h:13]: please remove all uses of `NS_ASSUME_NONNULL_BEGIN` from the PR and replace with individual nullability annotations.

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h
Comment (iskakaushik) [inline shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h:32]: Given that this is exposing new public API. I'd like @stuartmorgan to LGTM this PR before we land it.

**Comment by chinmaygarde** [comment]
@stuartmorgan is out till Monday. Let's wait for a bit for his review.

**Comment by stuartmorgan** [comment]
Sorry for the delay; I was OOO for most of December. I'll take a look today or tomorrow.

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h:18]: Calling an object *View when it's not an NSView is a confusing violation of convention. If it's to hold a view, call it something like ...ViewWrapper.

I'm a little confused as to what the purpose of this protocol is though. If it's only ever supposed to return one view (per the comments below), and its only method is to access the view... why not just have the view factory return a view? What problem does this intermediate class solve?

(I see that the iOS code has this too, and generally I'm in favor of matching APIs, but I'd like to understand what purpose it serves before deciding whether to mirror it.)

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h:23]: "Undefined behavior"? That's extremely bad if true in the technical sense of the term. What's the reason for that?

Also, this comment makes it sound like making and caching it on demand in this method isn't allowed, which I assume isn't actually the case?

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h
Comment (stuartmorgan) [reply shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h:32]: Maybe I'm missing something in the PR; where is the public API? I don't see any new public headers.

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h:34]: Nit: Creates

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h:36]: Nit: [...] macOS [...] exposes [...]

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h:40]: Related to my confusion about what FlutterPlatformView is/does, I'm not sure what it means for a FlutterPlatformView to have a size, since it's not itself a view.

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h:49]: nit: `nullable id`

Applies to all uses of `_Nullable` in the PR, since they are all in methods, where `nullable` works.

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViewController.mm
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterPlatformViewController.mm:39]: Why is this always nil?

File: shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Source/FlutterPlatformViews.h:57]: This name is really unfortunate; it sounds like a `create*` method, which it's not, and `args` shouldn't be abbreviated. I guess we're stuck with this due to the iOS naming though.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
I did a first pass. I'm confused by what's expected to work as a result of this PR though, given that there are references to adding public API but this PR doesn't appear to actually add any public API.

**Comment by RichardJCai** [comment]
> I did a first pass. I'm confused by what's expected to work as a result of this PR though, given that there are references to adding public API but this PR doesn't appear to actually add any public API.

Thanks for the review! I'll update this on Friday.

**Comment by RichardJCai** [comment]
@stuartmorgan @iskakaushik 
I'm trying to make progress on this but for some reason, whenever I try building with a local engine, I'm getting the following error.

```
Building macOS application...
2021-01-18 16:42:24.578 testapp3[52826:2671766] Failed to find path for "flutter_assets"
2021-01-18 16:42:24.578 testapp3[52826:2671766] Failed to find path for "flutter_assets"
embedder.cc (764): 'FlutterEngineInitialize' returned 'kInvalidArguments'. The assets path in the Flutter project arguments was missing.
2021-01-18 16:42:24.578 testapp3[52826:2671766] Failed to initialize Flutter engine: error 2
```

First ran into after rebasing this branch ontop of master but it seems like I'm running into on master as well with a local engine.

Tried following the suggestions in this issue https://github.com/flutter/flutter/issues/72608, ie upgrading and doing a new project but none seem to work with a local engine.

`flutter run -d macos` seems to work fine.

Anyone have any ideas whats wrong here?

**Comment by stuartmorgan** [comment]
I wouldn't have expected a local engine to have any effect on App.framework copying. I'll see if I can reproduce locally and follow up in https://github.com/flutter/flutter/issues/72608

**Comment by jmagman** [comment]
I'm getting the same failures as https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8857625998621949056/+/steps/build_host_debug_unopt/0/stdout locally.  Can you fix those up and see if you are still reproducing https://github.com/flutter/flutter/issues/72608?

**Comment by RichardJCai** [comment]
> I'm getting the same failures as https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8857625998621949056/+/steps/build_host_debug_unopt/0/stdout locally. Can you fix those up and see if you are still reproducing [flutter/flutter#72608](https://github.com/flutter/flutter/issues/72608)?

Never mind, I was able to solve the problem by recloning the repos and setting up the environment again.

**Comment by iskakaushik** [comment]
Hey @RichardJCai , thanks for continuing to work and make progress on this PR. Given the inactivity over the last week, I'm closing this. Feel free to open a new PR once you get a chance to address the failing pre-submit checks and tag me. Thanks!

**Comment by RichardJCai** [comment]
> Hey @RichardJCai , thanks for continuing to work and make progress on this PR. Given the inactivity over the last week, I'm closing this. Feel free to open a new PR once you get a chance to address the failing pre-submit checks and tag me. Thanks!

Sorry tried to address the PR comments but I still can't seem to get my engine environment set up for some reason.

**Comment by cbenhagen** [comment]
@RichardJCai did you have a chance to fix your environment? I know you probably already checked this but I once got this exact same error message when the underlying issue was in fact a build error in my dart code. Can you maybe post a verbose output of your build with the local engine? I'd be happy to help you debug your environment as good as I can. Here on or on Discord (cben#2483).

**Pull request closed without merging**
